### PR TITLE
fix: Rewrite ptytest to buffer stdout

### LINKED ISSUE #2122: Tests using Cobra with `ptytest` can hang on macOS (and Windows) unless the output is consumed
Unless `ptytest` output is consumed, a test can hang if a command outputs more than `1024` bytes (at least on macOS).

There are two ways to work around this issue:
1. Consume the output via `pty.ExpectMatch()` (ensure no more than 1024 bytes are left unread)
2. Use `cmd.SetOut(io.Discard)` instead

Since using `ptytest` is a common pattern in our tests, it'd be a good idea to rewrite it so that a hang is not possible.

Unless there's a bug to be fixed in underlying libraries, or changes to how ptys are configured, the most robust way to fix this would be for `ptytest` to buffer/consume _all_ command output, irregardless of a call to `ExpectMatch`.

A reproduction of this bug exists as a test in https://github.com/coder/coder/pull/1629 (and will be merged in and skipped on macOS and Windows).

### PR DESCRIPTION
**Pull request by mafredri**
With this PR we now always buffer command output.

The `ExpectMatch` function was also refactored a bit to avoid a race where a timeout could fire whilst writing to buffer (concurrent read/write), or a timeout could fail the test just as it succeeds.

Fixes #2122

### CODE REVIEW

File: pty/ptytest/ptytest.go [hunk 1]
```diff
@@ -7,116 +7,247 @@ import (
...
 	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
 
 	"github.com/coder/coder/pty"
 )
 
-var (
-	// Used to ensure terminal output doesn't have anything crazy!
-	// See: https://stackoverflow.com/a/29497680
-	stripAnsi = regexp.MustCompile("[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))")
```
Comment (mafredri) [inline pty/ptytest/ptytest.go:25]: Not needed because we output all lines with `%q`.

File: pty/ptytest/ptytest.go [hunk 2]
```diff
@@ -7,116 +7,248 @@ import (
...
+
+	timeout := make(chan error, 1)
 	go func() {
+		defer close(timeout)
 		timer := time.NewTimer(10 * time.Second)
 		defer timer.Stop()
 		select {
 		case <-complete.Done():
 			return
 		case <-timer.C:
```
Comment (kylecarbs) [inline pty/ptytest/ptytest.go:100]: Could we select this channel in the block below? Seems like we could reduce code!

**Review by kylecarbs — APPROVED** [review]
Looks good to me! Awesome change that should reduce a LOT of potential races. 😎

File: pty/ptytest/ptytest.go (hunk 2 above)
Comment (mafredri) [reply pty/ptytest/ptytest.go:100]: Doh, of course, that's such a good idea, why didn't I see it :joy:.

**Pull request merged**